src/components/controller.cpp

```diff
@@ -76,42 +76,33 @@ controller::controller(connection& conn, signal_emitter& emitter, const logger&
 
   m_log.trace("controller: Setup user-defined modules");
   size_t created_modules{0};
+  created_modules += setup_modules(alignment::LEFT);
+  created_modules += setup_modules(alignment::CENTER);
+  created_modules += setup_modules(alignment::RIGHT);
 
-  for (int i = 0; i < 3; i++) {
-    alignment align{static_cast<alignment>(i + 1)};
-    string configured_modules;
-
-    if (align == alignment::LEFT) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-left", ""s);
-    } else if (align == alignment::CENTER) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-center", ""s);
-    } else if (align == alignment::RIGHT) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-right", ""s);
-    }
-
-    for (auto& module_name : string_util::split(configured_modules, ' ')) {
-      if (module_name.empty()) {
-        continue;
-      }
-
-      try {
-        auto type = m_conf.get("module/" + module_name, "type");
-
-        if (type == "custom/ipc" && !m_ipc) {
-          throw application_error("Inter-process messaging needs to be enabled");
-        }
+  if (!created_modules) {
+    throw application_error("No modules created");
+  }
 
-        m_modules[align].emplace_back(make_module(move(type), m_bar->settings(), module_name, m_log));
-        created_modules++;
-      } catch (const runtime_error& err) {
-        m_log.err("Disabling module \"%s\" (reason: %s)", module_name, err.what());
+  /*
+   * Check if each module name only appears once
+   * Iterate over each module and check if its name appears anywhere else
+   */
+  for (auto it = m_modules.begin(); it != m_modules.end(); it++) {
+    auto name = it->get()->name();
```

Comment:
```suggestion
    const auto& name = it->get()->name();
```

Avoid copying the name each time.

EDIT:
The method `virtual string name() const = 0;` doesn't return a const ref but definitively should.